torch/distributed/pipeline/sync/checkpoint.py

```diff
@@ -99,31 +100,31 @@ def __init__(self, function: Function, batch: Batch) -> None:
     def checkpoint(self) -> Batch:
         """Returns a batch applied by :class:`Checkpoint`."""
         input_atomic = self.batch.atomic
-        input = tuple(self.batch)
+        inputs = tuple(self.batch)
 
         # Use a phony which requires grad to ensure that Checkpoint can be
         # tracked by the autograd engine even when none of the input tensors
         # require grad.
-        phony = get_phony(self.batch[0].device, requires_grad=True)
+        phony = get_phony(self.batch.get_device(), requires_grad=True)
 
-        output = Checkpoint.apply(phony, self.recomputed, self.rng_states, self.function, input_atomic, *input)
+        output = Checkpoint.apply(phony, self.recomputed, self.rng_states, self.function, input_atomic, *inputs)
 
         # Gradients are only supported for float Tensors.
         if isinstance(output, tuple):
-            output = tuple([x if x.is_floating_point() else x.detach() for x in output])
+            output = tuple([x.detach() if torch.is_tensor(x) and not x.is_floating_point() else x for x in output])
 
         return Batch(output)
 
     def recompute(self, batch: Batch) -> None:
         """Applies :class:`Recompute` to the batch in place."""
         input_atomic = self.batch.atomic
-        input = tuple(self.batch)
+        inputs = tuple(self.batch)
 
-        # batch[0] is always requiring grad, because it has been passed
```

Comment:
Shall we keep the comments here? Only need to change `batch[0]` to `batch[tensor_idx]` I guess.